sirepo/cookie.py

```diff
@@ -0,0 +1,131 @@
+# -*- coding: utf-8 -*-
+u"""User session management via an HTTP cookie
+
+:copyright: Copyright (c) 2015 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from __future__ import absolute_import, division, print_function
+
+from pykern import pkcollections
+from pykern import pkconfig
+from pykern.pkdebug import pkdc, pkdexc, pkdlog, pkdp
+from sirepo import util
+import base64
+import flask
+import re
+import sirepo.beaker_compat
+
+# Max age in seconds
+_COOKIE_MAX_AGE = 10 * 365 * 24 * 3600
+
+#: Identifies if the session has been returned by the client
+_SESSION_KEY_COOKIE_SENTINEL = 'srk'
+
+#: Identifies the user in the session
+_SESSION_KEY_USER = 'sru'
+
+
+def clear_user():
+    if has_key(_SESSION_KEY_USER):
+        remove(_SESSION_KEY_USER)
+
+
+def get(key):
+    return _cookie()[key]
+
+
+def get_user(checked=True):
+    return _cookie().get_user(checked)
+
+
+def has_key(key):
+    return key in _cookie()
+
+
+def init(logger):
+    header = flask.request.environ['HTTP_COOKIE'] if 'HTTP_COOKIE' in flask.request.environ else ''
+    _cookie(_Session(header, logger))
+
+
+def remove(key):
+    del _cookie()[key]
+
+
+def save_to_cookie(response):
+    _cookie().save_to_cookie(response)
+
+
+def set(key, value):
+    _cookie()[key] = value
+
+
+def set_user(uid):
+    _cookie().set_user(uid)
+
+
+def init_mock(uid):
+    """A mock session for pkcli"""
+    res = _Session('', None)
+    res[_SESSION_KEY_USER] = uid
+    res[_SESSION_KEY_COOKIE_SENTINEL] = 1
+    flask.g = pkcollections.Dict({
+        'sirepo_cookie': res,
+    })
+
+
+def _cookie(v=None):
+    if v is not None:
+        flask.g.sirepo_cookie = v
+    return flask.g.sirepo_cookie
+
+
+class _Session(dict):
+
+    def __init__(self, header, logger):
+        self.logger = logger
+        self._from_cookie_header(header)
+        self._log_user()
+
+    def get_user(self, checked=True):
+        if not self.get(_SESSION_KEY_COOKIE_SENTINEL):
+            util.raise_forbidden('Missing session, cookies may be disabled')
+        return self[_SESSION_KEY_USER] if checked else self.get(_SESSION_KEY_USER)
+
+    def save_to_cookie(self, response):
+        if 200 <= response.status_code < 400:
+            self[_SESSION_KEY_COOKIE_SENTINEL] = 1
+            #TODO(pjm): don't set if it already matches
```

Comment:
```py
if _MODIFIED_KEY in self:
    del self[_MODIFIED_KEY]
    response.set_cookie
```

Reply:
TODO: uri_router needs a /static route and static_files needs to be false in server.